### DISCUSSION TITLE
Extending API and Web-App for Custom Service

### DISCUSSION THREAD

**Discussion by jbell-astrobotic** [yamcs/yamcs#852] [Q&A]
Hi! Reading through the CHANGELOG for version 5.8.6 and I see "First prototype for web extensions to allow plugins to add extensions to yamcs-web". Looking for additional information on the current scope/functionality of this work (as well as maybe any additional info on future plans). Also, I'm very interested in testing this out - looking to get some basic sample code or at least just additional information to point me in the right direction to get a sample up and running.
Also, from reading this -> https://docs.yamcs.org/yamcs-server-manual/yamcs-plugin-format/ -- it says "maybe you want to add your own custom HTTP and WebSocket calls to the API", and that is something I would like to do, but unable to figure out the actual implementation for adding endpoints to the API without creating a custom HttpServer. Again, looking for either sample code or a push in the right direction to obtain further information on how to implement this.

Ultimately, I'd like to create a custom service, extend the REST API to include additional GET and POST calls for this custom service, and have some frontend widgets in the web app.

Let me know where I can clarify or if what I'm requesting isn't clear!!

**Comment by fqqb** [comment]
For custom HTTP methods, you can look at yamcs-web as an example, this module is implemented as a Yamcs plugin, and adds one extra method specified through a GPB service: https://github.com/yamcs/yamcs/blob/master/yamcs-web/src/main/proto/yamcs/protobuf/web/web.proto

Custom widgets is not something we currently provide any help on, because this is very beta. It is working in some non-public extensions, but needs more polishing. We use it to add custom pages (not widgets) in the sidebar of a yamcs-web instance. The alternative, I believe, is to make a custom web application.
